transaction_manager.py

```diff
@@ -53,9 +56,18 @@ def read(
         available_sites = self.site_manager.get_available_sites(data_id)
         read_ready_sites = [site for site in available_sites if site in previously_running_sites]
 
+        # Move the transaction to the waiting set
         if not read_ready_sites:
             print(f"No sites available - Moving (R,{t_id},{data_id}) to pending reads")
+            self.waiting_set[t_id] = self.waiting_set.get(t_id, 0) + 1
+            for site_id in previously_running_sites:
+                self.site_manager.add_to_pending_reads(site_id, t_id, data_id)
+            return
 
+        # Do not process a waiting transaction
+        if t_id in self.waiting_set:
```

Comment:
We can add a criteria to skip this check if is_pending_read is True
```suggestion
        if t_id in self.waiting_set and not is_pending_read:
....
```